cmd/ecc/buffer.go

```diff
@@ -0,0 +1,167 @@
+/*
+ * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ecc
+
+// The Buffer implementation below is tightly coupled to
+// the JoinedReaders implementation in this package.
+
+// Buffer contains a set of data and parity shards.
+// It translates between a flat memory and multi-dimensional
+// slices (shards).
+type Buffer struct {
+	shards [][]byte
+	data   [][]byte // view of shards containing actual data
+	parity [][]byte // view of shards containing parity data
+
+	buffer []byte
+	offset int
```

Comment:
Instead of keeping the `offset` you could consider `buffer` to contain the data remaining of the current shard and update the slice as it is read. 

This means that data remaining is data in `buffer` plus anything in `data`, making `offset` obsolete. I am pretty sure that would simplify the code.

Reply:
I did a similar thing already and removed the buffer